browser/ui/views/brave_rewards/tip_panel_bubble_host.cc

```diff
@@ -0,0 +1,144 @@
+  }
+
+  // views::Widget:
+  std::unique_ptr<views::NonClientFrameView> CreateNonClientFrameView(
+      views::Widget* widget) override {
+    auto frame = WebUIBubbleDialogView::CreateNonClientFrameView(widget);
+    auto* bubble_frame = static_cast<views::BubbleFrameView*>(frame.get());
+    auto* bubble_border = bubble_frame->bubble_border();
+    DCHECK(bubble_border);
+    bubble_border->SetRoundedCorners(0, 0, 16, 16);
+    return frame;
+  }
+
+  // views::View:
+  void AddedToWidget() override {
+    WebUIBubbleDialogView::AddedToWidget();
+    web_view()->holder()->SetCornerRadii(gfx::RoundedCornersF(0, 0, 16, 16));
+  }
+};
+
+// In order to use `TipPanelDialogView` as the dialog view implementation, we
+// must override the `CreateWebUIBubbleDialog` of `WebUIBubbleManager`.
+class TipPanelBubbleManager : public WebUIBubbleManager {
+ public:
+  TipPanelBubbleManager(views::View* anchor_view, Profile* profile)
+      : anchor_view_(anchor_view), profile_(profile) {}
+
+  ~TipPanelBubbleManager() override = default;
+
+  // WebUIBubbleManager:
+  void MaybeInitPersistentRenderer() override {}
+
+  base::WeakPtr<WebUIBubbleDialogView> CreateWebUIBubbleDialog(
+      const absl::optional<gfx::Rect>& anchor) override {
+    auto contents_wrapper =
+        std::make_unique<BubbleContentsWrapperT<TipPanelUI>>(
+            GURL(kBraveTipPanelURL), profile_, IDS_BRAVE_UI_BRAVE_REWARDS);
+
+    set_bubble_using_cached_web_contents(false);
+    set_cached_contents_wrapper(std::move(contents_wrapper));
+    cached_contents_wrapper()->ReloadWebContents();
+
+    auto bubble_view = std::make_unique<TipPanelDialogView>(
+        anchor_view_, cached_contents_wrapper(), anchor);
+
+    auto weak_ptr = bubble_view->GetWeakPtr();
+    views::BubbleDialogDelegateView::CreateBubble(std::move(bubble_view));
+    return weak_ptr;
+  }
+
+ private:
+  const raw_ptr<views::View> anchor_view_;
+  const raw_ptr<Profile> profile_;
+};
+
+views::View* GetAnchorView(Browser* browser) {
+  DCHECK(browser);
+  auto* browser_view = BrowserView::GetBrowserViewForBrowser(browser);
+  DCHECK(browser_view);
+  auto* location_bar_view = browser_view->GetLocationBarView();
+  DCHECK(location_bar_view);
+  return location_bar_view;
+}
+
+}  // namespace
+
+TipPanelBubbleHost::TipPanelBubbleHost(Browser* browser)
+    : BrowserUserData<TipPanelBubbleHost>(*browser) {
+  auto* coordinator = TipPanelCoordinator::FromBrowser(browser);
+  DCHECK(coordinator);
+  coordinator_observation_.Observe(coordinator);
+}
+
+TipPanelBubbleHost::~TipPanelBubbleHost() = default;
+
+void TipPanelBubbleHost::MaybeCreateForBrowser(Browser* browser) {
+  DCHECK(browser);
+  if (IsSupportedForProfile(browser->profile())) {
+    CreateForBrowser(browser);
+  }
+}
+
+void TipPanelBubbleHost::OnTipPanelRequested(const std::string& publisher_id) {
+  // No-op if the bubble is already open.
+  if (bubble_manager_ && bubble_manager_->GetBubbleWidget()) {
```

Comment:
Do we need to verify that if the bubble is already open that it's for the specified `publisher_id`? Or is that a given at this point?

Reply:
I'll add a comment here. It seems reasonable to no-op if the panel is open with a different publisher displayed, so that the user can continue the UI flow for that publisher.